src/backend/cuda/culibs/CMakeLists.txt

```diff
@@ -0,0 +1,110 @@
+# Copyright (c) 2020, ArrayFire
+# All rights reserved.
+#
+# This file is distributed under 3-clause BSD license.
+# The complete license agreement can be obtained at:
+# http://arrayfire.com/licenses/BSD-3-Clause
+
+if(UNIX)
+  # Static linking cuda libs require device linking, which in turn
+  # requires separable compilation.
+  set(pior_val_CUDA_SEPARABLE_COMPILATION OFF)
+  if(DEFINED CUDA_SEPARABLE_COMPILATION)
+    set(pior_val_CUDA_SEPARABLE_COMPILATION ${CUDA_SEPARABLE_COMPILATION})
+  endif()
+  set(CUDA_SEPARABLE_COMPILATION ON)
+
+  # Find if CUDA Toolkit is at least 10.0 to use static
+  # lapack library. Otherwise, we have to use regular shared library
+  if(CUDA_VERSION_MAJOR VERSION_GREATER 10 OR CUDA_VERSION_MAJOR VERSION_EQUAL 10)
+    set(AF_USE_STATIC_CUDA_LAPACK ON CACHE INTERNAL "")
+  else()
+    set(AF_USE_STATIC_CUDA_LAPACK OFF CACHE INTERNAL "")
+  endif()
+
+  # The following macro uses a macro defined by
+  # FindCUDA module from cmake.
+  macro(af_find_static_cuda_libs libname)
+  set(search_name
+     "${CMAKE_STATIC_LIBRARY_PREFIX}${libname}${CMAKE_STATIC_LIBRARY_SUFFIX}")
+  cuda_find_library_local_first(CUDA_${libname}_LIBRARY
+      ${search_name} "${libname} static library")
+  mark_as_advanced(CUDA_${libname}_LIBRARY)
+  endmacro()
+
+  af_find_static_cuda_libs(culibos)
+  af_find_static_cuda_libs(cublas_static)
+  af_find_static_cuda_libs(cublasLt_static)
+  af_find_static_cuda_libs(cufft_static)
+  af_find_static_cuda_libs(cusparse_static)
+
+  set(extra_nvcc_flags "-rdc=true;-dlink;-lculibos")
+  set(extra_nvcc_flags "${extra_nvcc_flags};-lcublas_static;-lcublasLt_static")
+  set(extra_nvcc_flags "${extra_nvcc_flags};-lcufft_static;-lcusparse_static")
+
+  if(${AF_USE_STATIC_CUDA_LAPACK})
+    af_find_static_cuda_libs(cusolver_static)
+
+    # NVIDIA LAPACK library liblapack_static.a is a subset of LAPACK and only
+    # contains GPU accelerated stedc and bdsqr. The user has to link
+    # libcusolver_static.a with liblapack_static.a in order to build the
+    # application successfully.
+    af_find_static_cuda_libs(lapack_static)
+
+    set(extra_nvcc_flags "${extra_nvcc_flags};-lcusolver_static")
+  endif()
+endif()
+
+cuda_add_library(cuda_culibs STATIC
+    blas.cu
+    blas.hpp
+    cufft.cu
+    cufft.hpp
+    fft.cu
+    sparse.cu
+    sparse.hpp
+    sparse_arith.cu
+    sparse_arith.hpp
+    sparse_blas.cu
+    sparse_blas.hpp
+    solve.cu
+    solve.hpp
```

Comment:
There is no context why these files are used in this cmake file. I don't like this approach. If you want to keep the cmake separate add this file to the CMakeModules folder in the root arrayfire directory create a function that adds these files with an explanation.

Reply:
I am not saying what I have now is optimum approach, but putting it in CMakeModules isn't correct. This section of code is very specific to CUDA backend and much like how we had `cuda_thrust_sort_by_key` and `cuda_scan_by_key` targets, this also is including sources from `src/backend/cuda`. Therefore, I think this file should be under `src/backend/cuda`. I have no specific opinion as to where under `src/backend/cuda`, anywhere would work as long as we `include` it properly.